app/api/discount_codes.py

```diff
@@ -3,47 +3,69 @@
+        :param json_data:
+        :return:"""
+
+        used_for = json_data['data']['attributes'].get('used-for')
+        if used_for in ('event', 'ticket'):
+            if used_for == 'event':
+                self.schema = DiscountCodeSchemaEvent
+            elif used_for == 'ticket':
+                self.schema = DiscountCodeSchemaTicket
+        else:
+            raise ConflictException(
+                {'pointer': '/data/attributes/used-for'},
+                "used-for attribute is required and should be equal to 'ticket' or 'event' to create discount code")
+
     def before_post(self, args, kwargs, data):
+        """Before post method to check required relationships and set user_id
+        :param args:
+        :param kwargs:
+        :param data:
+        :return:"""
         if data['used_for'] == 'ticket':
-            self.schema = DiscountCodeSchemaTicket
             require_relationship(['event'], data)
             if not has_access('is_coorganizer', event_id=data['event']):
                 raise ForbiddenException({'source': ''}, 'You are not authorized')
-        elif data['used_for'] == 'event' and has_access('is_admin') and 'events' in data:
-            self.schema = DiscountCodeSchemaEvent
-        else:
+        elif not data['used_for'] == 'event' and has_access('is_admin') and 'events' in data:
             raise UnprocessableEntity({'source': ''}, "Please verify your permission or check your relationship")
 
         data['user_id'] = current_identity.id
 
     def before_create_object(self, data, view_kwargs):
-        if data['used_for'] == 'event' and 'events' in data:
-            for event in data['events']:
-                try:
-                    event_now = db.session.query(Event).filter_by(id=event, deleted_at=None).one()
-                except NoResultFound:
-                    raise UnprocessableEntity({'event_id': event},
-                                              "Event does not exist")
-                if event_now.discount_code_id:
-                    raise UnprocessableEntity({'event_id': event},
-                                              "A Discount Code already exists for the provided Event ID")
+        if data['used_for'] == 'event':
+            self.resource.schema = DiscountCodeSchemaEvent
+            if 'events' in data:
+                for event in data['events']:
+                    try:
+                        event_now = db.session.query(Event).filter_by(id=event, deleted_at=None).one()
+                    except NoResultFound:
+                        raise UnprocessableEntity({'event_id': event}, "Event does not exist")
+                    if event_now.discount_code_id:
+                        raise UnprocessableEntity(
+                            {'event_id': event}, "A Discount Code already exists for the provided Event ID")
+        else:
+            self.resource.schema = DiscountCodeSchemaTicket
```

Comment:
Why do we need this now that the schema is already decided in the `decide_schema` method ?

Reply:
We need this because `decide_schema` is used for changing `self.schema` only. Ticket relationship is not registered without correct `self.resource.schema` so I added it.

Reply:
Got it. :+1: